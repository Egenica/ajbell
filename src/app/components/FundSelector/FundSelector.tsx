import React, { useEffect, useState } from 'react';

type FundSelectorProps = {
  onSelectFund: (fund: string) => void;
};

const FundSelector: React.FC<FundSelectorProps> = ({ onSelectFund }) => {
  const [strategy, setStrategy] = useState<string | null>(null);
  const [selectedFund, setSelectedFund] = useState<string>('');

  const growthFunds = ['Cautious', 'Balanced', 'Adventurous'];
  const responsibleFund = ['Responsible'];

  useEffect(() => {
    const savedStrategy = localStorage.getItem('strategy');
    const savedFund = localStorage.getItem('selectedFund');
    if (savedStrategy) setStrategy(savedStrategy);
    if (savedFund) {
      setSelectedFund(savedFund);
      onSelectFund(savedFund);
    }
  }, [onSelectFund]);

  const handleStrategyChange = (e: React.ChangeEvent<HTMLSelectElement>) => {
    const newStrategy = e.target.value;
    setStrategy(newStrategy);
    localStorage.setItem('strategy', newStrategy);
    setSelectedFund('');
    onSelectFund(''); // Reset fund selection on strategy change
    localStorage.removeItem('selectedFund'); // Clear selected fund from local storage
  };

  const handleFundChange = (e: React.ChangeEvent<HTMLSelectElement>) => {
    const newFund = e.target.value;
    setSelectedFund(newFund);
    localStorage.setItem('selectedFund', newFund);
    onSelectFund(newFund);
  };

  return (
    <div className="p-6 border rounded-lg bg-white shadow-lg">
      <h2 className="text-2xl font-bold mb-4">Choose a Strategy</h2>
      <div className="mb-4">
        <label className="block mb-2">
          Select Strategy:
          <select
            className="w-full p-2 border rounded"
            value={strategy || ''}
            onChange={handleStrategyChange}
          >
            <option value="">-- Select Strategy --</option>
            <option value="growth">Growth Funds</option>
            <option value="responsible">Responsible Fund</option>
          </select>
        </label>
      </div>

      {strategy === 'growth' && (
        <div className="mb-4">
          <label className="block mb-2">
            Select Growth Fund:
            <select
              className="w-full p-2 border rounded"
              value={selectedFund}
              onChange={handleFundChange}
            >
              <option value="">-- Select Fund --</option>
              {growthFunds.map((fund) => (
                <option key={fund} value={fund.toLowerCase()}>
                  {fund}
                </option>
              ))}
            </select>
          </label>
        </div>
      )}

      {strategy === 'responsible' && (
        <div className="mb-4">
          <label className="block mb-2">Select Responsible Fund:</label>
          <select
            className="w-full p-2 border rounded"
            value={selectedFund}
            onChange={handleFundChange}
          >
            <option value="">-- Select Fund --</option>
            {responsibleFund.map((fund) => (
              <option key={fund} value={fund.toLowerCase()}>
                {fund}
              </option>
            ))}
          </select>
        </div>
      )}
    </div>
  );
};

export default FundSelector;
