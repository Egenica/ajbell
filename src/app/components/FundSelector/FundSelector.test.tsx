import React from 'react';
import { render, fireEvent } from '@testing-library/react';
import '@testing-library/jest-dom';
import FundSelector from './FundSelector';

describe('FundSelector Component', () => {
  it('should render correctly', () => {
    const { container } = render(<FundSelector onSelectFund={() => {}} />);
    expect(container).toMatchSnapshot();
  });

  it('should display growth funds when growth strategy is selected', () => {
    const { getByLabelText, getByText } = render(
      <FundSelector onSelectFund={() => {}} />
    );

    fireEvent.change(getByLabelText('Select Strategy:'), {
      target: { value: 'growth' },
    });

    expect(getByText('Select Growth Fund:')).toBeInTheDocument();
    expect(getByText('Cautious')).toBeInTheDocument();
    expect(getByText('Balanced')).toBeInTheDocument();
    expect(getByText('Adventurous')).toBeInTheDocument();
  });

  it('should display responsible fund when responsible strategy is selected', () => {
    const { getByLabelText, getByText } = render(
      <FundSelector onSelectFund={() => {}} />
    );

    fireEvent.change(getByLabelText('Select Strategy:'), {
      target: { value: 'responsible' },
    });

    expect(getByText('Select Responsible Fund:')).toBeInTheDocument();
    expect(getByText('Responsible')).toBeInTheDocument();
  });

  it('should call onSelectFund with the correct value when a fund is selected', () => {
    const onSelectFundMock = jest.fn();
    const { getByLabelText } = render(
      <FundSelector onSelectFund={onSelectFundMock} />
    );

    fireEvent.change(getByLabelText('Select Strategy:'), {
      target: { value: 'growth' },
    });
    fireEvent.change(getByLabelText('Select Growth Fund:'), {
      target: { value: 'cautious' },
    });

    expect(onSelectFundMock).toHaveBeenCalledWith('cautious');
  });

  it('should reset fund selection when strategy changes', () => {
    const onSelectFundMock = jest.fn();
    const { getByLabelText } = render(
      <FundSelector onSelectFund={onSelectFundMock} />
    );

    fireEvent.change(getByLabelText('Select Strategy:'), {
      target: { value: 'growth' },
    });
    fireEvent.change(getByLabelText('Select Growth Fund:'), {
      target: { value: 'cautious' },
    });
    fireEvent.change(getByLabelText('Select Strategy:'), {
      target: { value: 'responsible' },
    });

    expect(onSelectFundMock).toHaveBeenCalledWith('');
  });
});
