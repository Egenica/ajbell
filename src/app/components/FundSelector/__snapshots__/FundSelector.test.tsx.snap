// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FundSelector Component should render correctly 1`] = `
<div>
  <div
    class="p-6 border rounded-lg bg-white shadow-lg md:sticky md:top-6"
  >
    <h2
      class="text-2xl font-bold mb-4"
    >
      Choose a Strategy
    </h2>
    <div
      class="mb-4"
    >
      <label
        class="block mb-2"
      >
        Select Strategy:
        <select
          class="w-full p-2 border rounded"
        >
          <option
            value=""
          >
            -- Select Strategy --
          </option>
          <option
            value="growth"
          >
            Growth Funds
          </option>
          <option
            value="responsible"
          >
            Responsible Fund
          </option>
        </select>
      </label>
    </div>
  </div>
</div>
`;
